src/generic/maskededit.cpp

```diff
@@ -0,0 +1,1951 @@
+    int kc = 0;
+#if wxUSE_UNICODE
+    kc = (int) event.GetUnicodeKey();
+#else // !wxUSE_UNICODE
+    kc = event.GetKeyCode();
+#endif // wxUSE_UNICODE/!wxUSE_UNICODE
+
+    //Allow default 'ctrl-C' manager
+    if ( kc == WXK_CONTROL_C)
+        return;
+
+    //Allow standard 32-127 ASCII and also 128-255 ASCII extended characters
+    if ( kc < WXK_SPACE || kc >= WXK_START )
+        return;
+
+    wxChar keyChar(kc);
+
+    event.Skip(false); //This event will end here.
+
+    //If an EVT_CHAR event was generated for these cases, don't process
+    //them again. They have been already processed at OnKeyDown
+    switch ( (int)keyChar )
+    {
+        case WXK_LEFT:
+        case WXK_NUMPAD_LEFT:
+        case WXK_RIGHT:
+        case WXK_NUMPAD_RIGHT:
+        case WXK_BACK:
+        case WXK_DELETE:
+        case WXK_NUMPAD_DELETE:
+            return;
+
+        default:
+            break;
+    }
+
+    long caretPos = GetTextEntry()->GetInsertionPoint();
+
+    m_fillBck = m_filled;
+    //First handle the selection, if any
+    long selPos = -1; //changes if a selection erases anything
+    bool needsUpdate = false;
+    //The whole control's value, changed after deleting the selection;
+    // and the new caret position.
+    wxString workString = HandleSelection(&selPos, &needsUpdate);
+    if ( selPos != -1 )
+        caretPos = selPos;
+
+    bool isProccesed = ProcessChar(keyChar, &caretPos, &workString, &needsUpdate);
+
+    //Update if necessary
+    if ( isProccesed )
+    {
+        //If there was no selection and the typed char is a literal, we only
+        // move the caret, without useless SetValue().
+        //And avoid so the false wxEVT_COMMAND_TEXT_UPDATED event.
+        if ( needsUpdate || selPos != -1)
+            GetTextEntry()->SetValue(workString);
+
+        MoveCaret(caretPos, true);
+    }
+    else
+    {
+        //restore
+        m_filled = m_fillBck;
+        //beep
+        DoBell();
+    }
+}
+
+void wxMaskedEdit::OnCut(wxCommandEvent& event)
+{
+    event.Skip(false);
+    DoCut();
+}
+
+void wxMaskedEdit::OnPaste(wxCommandEvent& event)
+{
+    event.Skip(false);
+    DoPaste();
+}
+
+#if wxUSE_COMBOBOX
+void wxMaskedEdit::OnComboSelect(wxCommandEvent& event)
+{
+    //We must process this new value (test, tell empty cells, etc)
+    SetMaskedValue(event.GetString());
+    UpdateMaskedColours();
+    GetTextEntry()->SetInsertionPoint(0);
+
+    event.Skip(true);
+}
+
+void wxMaskedEdit::OnComboDrop(wxCommandEvent& event)
+{
+    //Neither OK nor Wr[ong] colours are good here. Use colours for an empty value
+    UseColours(mEmColour);
+
+    event.Skip(true);
+}
+#endif // wxUSE_COMBOBOX
+
+// ---------------------------------------------------------------------------
+// -  -  -  -  -  -  -  - More helpers -  -  -  -  -  -  -  -  -  -  -  -  -
+
+//Our control derives from wxTextEntry, but not this class. Get a pointer to it
+wxTextEntry* wxMaskedEdit::GetTextEntry() const
+{
```

Comment:
Are `wxDynamicCast`s here really necessary? What happens if `msk_control` is always returned?